File: stdlib/lua/hull/csv.lua
--- RFC 4180 CSV parser + writer.
--
-- @module hull.csv
-- @license AGPL-3.0-or-later
--
-- Character-by-character state machine parser. Correctly handles quoted
-- fields containing the separator, embedded newlines, and doubled-quote
-- escapes (`""`).

local csv = {}

-- States for the parser state machine
local STATE_FIELD_START  = 1
local STATE_UNQUOTED     = 2
local STATE_QUOTED        = 3
local STATE_QUOTE_IN_QUOTED = 4

--- Parse a CSV string.
--
-- @tparam string text  CSV text (UTF-8). `nil` or `""` returns `{}`.
-- @tparam[opt] table opts  Options:
--
--   - `headers`   (boolean, default `false`): first row treated as
--     headers; returns an array of objects (`{ [col_name] = value, ... }`)
--     instead of an array of arrays.
--   - `separator` (string, default `","`): field delimiter.
--   - `quote`     (string, default `'"'`): quote character.
--   - `max_rows`  (integer, default `100_000`): hard cap; rows past this
--     are silently dropped.
--
-- @treturn table[]  Array of row arrays (or row objects if `headers=true`).
-- @usage
-- local rows = csv.parse(req.body, { headers = true })
-- for _, row in ipairs(rows) do print(row.name, row.email) end
function csv.parse(text, opts)
    if text == nil or text == "" then
        return {}
    end

    opts = opts or {}
    local sep   = opts.separator or ","
    local quote = opts.quote or '"'
    local use_headers = opts.headers or false
    local max_rows = opts.max_rows or 100000

    local rows = {}
    local row = {}
    local field = {}
    local state = STATE_FIELD_START

    local len = #text
    local i = 1

    while i <= len do
        if #rows >= max_rows then
            error("csv.parse: exceeded max_rows limit (" .. max_rows .. ")")
        end
        local c = text:sub(i, i)

        if state == STATE_FIELD_START then
            if c == quote then
                -- Begin quoted field
                state = STATE_QUOTED
                i = i + 1
            elseif c == sep then
                -- Empty unquoted field, commit it
                row[#row + 1] = ""
                -- Stay in FIELD_START for next field
                i = i + 1
            elseif c == "\r" then
                -- End of row (CRLF or bare CR)
                row[#row + 1] = ""
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                if i + 1 <= len and text:sub(i + 1, i + 1) == "\n" then
                    i = i + 2
                else
                    i = i + 1
                end
            elseif c == "\n" then
                -- End of row (bare LF)
                row[#row + 1] = ""
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                i = i + 1
            else
                -- Begin unquoted field
                field[#field + 1] = c
                state = STATE_UNQUOTED
                i = i + 1
            end

        elseif state == STATE_UNQUOTED then
            if c == sep then
                -- End of field
                row[#row + 1] = table.concat(field)
                field = {}
                state = STATE_FIELD_START
                i = i + 1
            elseif c == "\r" then
                -- End of row (CRLF or bare CR)
                row[#row + 1] = table.concat(field)
                field = {}
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                if i + 1 <= len and text:sub(i + 1, i + 1) == "\n" then
                    i = i + 2
                else
                    i = i + 1
                end
            elseif c == "\n" then
                -- End of row (bare LF)
                row[#row + 1] = table.concat(field)
                field = {}
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                i = i + 1
            else
                field[#field + 1] = c
                i = i + 1
            end

        elseif state == STATE_QUOTED then
            if c == quote then
                -- Could be end of quoted field or escaped quote
                state = STATE_QUOTE_IN_QUOTED
                i = i + 1
            else
                -- Accumulate character (including newlines, separators, etc.)
                field[#field + 1] = c
                i = i + 1
            end

        elseif state == STATE_QUOTE_IN_QUOTED then
            if c == quote then
                -- Escaped quote (doubled)
                field[#field + 1] = quote
                state = STATE_QUOTED
                i = i + 1
            elseif c == sep then
                -- End of quoted field, then separator
                row[#row + 1] = table.concat(field)
                field = {}
                state = STATE_FIELD_START
                i = i + 1
            elseif c == "\r" then
                -- End of quoted field, then row end
                row[#row + 1] = table.concat(field)
                field = {}
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                if i + 1 <= len and text:sub(i + 1, i + 1) == "\n" then
                    i = i + 2
                else
                    i = i + 1
                end
            elseif c == "\n" then
                -- End of quoted field, then row end
                row[#row + 1] = table.concat(field)
                field = {}
                rows[#rows + 1] = row
                row = {}
                state = STATE_FIELD_START
                i = i + 1
            else
                -- Per RFC 4180 this is malformed, but be lenient:
                -- treat the closing quote as the end of quoting and
                -- continue accumulating as unquoted content
                field[#field + 1] = c
                state = STATE_UNQUOTED
                i = i + 1
            end
        end
    end

    -- Flush the last field/row if there is pending data
    if state == STATE_FIELD_START then
        -- We are at field start — only emit a row if we have accumulated
        -- fields (handles trailing separator case)
        if #row > 0 then
            row[#row + 1] = ""
            rows[#rows + 1] = row
        end
    elseif state == STATE_UNQUOTED then
        row[#row + 1] = table.concat(field)
        rows[#rows + 1] = row
    elseif state == STATE_QUOTED then
        -- Unterminated quoted field — flush what we have
        row[#row + 1] = table.concat(field)
        rows[#rows + 1] = row
    elseif state == STATE_QUOTE_IN_QUOTED then
        -- Closing quote was the last character
        row[#row + 1] = table.concat(field)
        rows[#rows + 1] = row
    end

    -- If headers mode, convert row arrays to objects
    if use_headers and #rows > 0 then
        local header_row = rows[1]
        local result = {}
        for r = 2, #rows do
            local obj = {}
            for c = 1, #header_row do
                obj[header_row[c]] = rows[r][c] or ""
            end
            result[#result + 1] = obj
        end
        return result
    end

    return rows
end

--- Encode rows into a CSV string.
--
-- @tparam table[] rows  Array of rows. When `opts.headers = true`, rows
--   are objects (keys become the header row). Otherwise rows are arrays.
-- @tparam[opt] table opts
--
--   - `headers`   (boolean, default `false`)
--   - `separator` (string, default `","`)
--   - `quote`     (string, default `'"'`)
--
-- @treturn string  CSV text with LF line endings. Values containing the
--   separator, quote, CR, or LF are auto-quoted; embedded quotes are
--   doubled (RFC 4180).
function csv.encode(rows, opts)
    if rows == nil or #rows == 0 then
        return ""
    end

    opts = opts or {}
    local sep   = opts.separator or ","
    local quote = opts.quote or '"'
    local use_headers = opts.headers or false
    local escaped_quote = quote .. quote

    -- Determine if a field value needs quoting
    local function needs_quoting(val)
        if val:find(sep, 1, true) then return true end
        if val:find(quote, 1, true) then return true end
        if val:find("\n", 1, true) then return true end
        if val:find("\r", 1, true) then return true end
        return false
    end

    -- Encode a single field
    local function encode_field(val)
        if val == nil then
            return ""
        end
        val = tostring(val)
        if needs_quoting(val) then
            return quote .. val:gsub(quote, escaped_quote) .. quote
        end
        return val
    end

    -- Encode a single row from an array of values
    local function encode_row(values)
        local fields = {}
        for i = 1, #values do
            fields[i] = encode_field(values[i])
        end
        return table.concat(fields, sep)
    end

    local lines = {}

    if use_headers then
        -- Collect header keys from the first row to establish column order
        local keys = {}
        if opts.columns then
            -- Allow explicit column ordering
            for i = 1, #opts.columns do
                keys[i] = opts.columns[i]
            end
        else
            -- Collect keys from all rows for consistent output
            local key_set = {}
            local key_order = {}
            for _, row in ipairs(rows) do
                for k, _ in pairs(row) do
                    if not key_set[k] then
                        key_set[k] = true
                        key_order[#key_order + 1] = k
                    end
                end
            end
            -- Sort for deterministic output
            table.sort(key_order)
            keys = key_order
        end

        -- Header row
        lines[#lines + 1] = encode_row(keys)

        -- Data rows
        for _, row in ipairs(rows) do
            local values = {}
            for i = 1, #keys do
                values[i] = row[keys[i]]
            end
            lines[#lines + 1] = encode_row(values)
        end
    else
        -- Array of arrays
        for _, row in ipairs(rows) do
            lines[#lines + 1] = encode_row(row)
        end
    end

    return table.concat(lines, "\n") .. "\n"
end

return csv

File: stdlib/js/hull/csv.js
/**
 * @file hull:csv
 * @module hull:csv
 * @description RFC 4180 CSV parser + writer. Lua parity: `hull.csv`.
 * @license AGPL-3.0-or-later
 */

/**
 * Parse a CSV string.
 *
 * @param {string} text  CSV text. Non-string or empty input returns `[]`.
 * @param {Object} [opts]
 * @param {boolean} [opts.headers=false]   First row treated as header; returns
 *   `Array<Object>` instead of `Array<Array>`.
 * @param {string}  [opts.separator=","]
 * @param {string}  [opts.quote='"']
 * @param {number}  [opts.maxRows=100000]  Lua-parity alias: `max_rows`.
 * @returns {Array<Array<string>>|Array<Object<string,string>>}
 */
function parse(text, opts) {
    if (typeof text !== "string" || text.length === 0) return [];

    const sep = (opts && opts.separator) || ",";
    const quo = (opts && opts.quote) || '"';
    const useHeaders = !!(opts && opts.headers);
    // L-3: accept Lua-style `max_rows` as well as the canonical
    // `maxRows`. `??` preserves an explicit 0 (the prior `||` would
    // have replaced it with the default).
    const maxRows = (opts && (opts.maxRows ?? opts.max_rows)) ?? 100000;

    const rows = [];
    let row = [];
    let field = "";
    let inQuoted = false;
    let i = 0;
    const len = text.length;

    while (i < len) {
        if (rows.length >= maxRows)
            throw new Error("csv.parse: exceeded maxRows limit (" + maxRows + ")");
        const ch = text[i];

        if (inQuoted) {
            if (ch === quo) {
                // Look ahead: doubled quote is an escape
                if (i + 1 < len && text[i + 1] === quo) {
                    field += quo;
                    i += 2;
                } else {
                    // End of quoted field
                    inQuoted = false;
                    i++;
                }
            } else {
                // Accumulate character (including newlines inside quotes)
                field += ch;
                i++;
            }
        } else {
            if (ch === quo && field.length === 0) {
                // Start of quoted field (quote must be at field start)
                inQuoted = true;
                i++;
            } else if (ch === sep) {
                // Field delimiter
                row.push(field);
                field = "";
                i++;
            } else if (ch === "\r") {
                // CRLF or bare CR -> end of row
                row.push(field);
                field = "";
                rows.push(row);
                row = [];
                i++;
                if (i < len && text[i] === "\n") i++;
            } else if (ch === "\n") {
                // LF -> end of row
                row.push(field);
                field = "";
                rows.push(row);
                row = [];
                i++;
            } else {
                field += ch;
                i++;
            }
        }
    }

    // Handle last field/row (no trailing newline case)
    // If there is any pending field content or the row has fields,
    // or the text did not end with a newline, push the final row.
    if (field.length > 0 || row.length > 0) {
        row.push(field);
        rows.push(row);
    }

    if (rows.length === 0) return [];

    if (useHeaders) {
        const headers = rows[0];
        const result = [];
        for (let r = 1; r < rows.length; r++) {
            const obj = {};
            for (let c = 0; c < headers.length; c++) {
                const key = headers[c];
                if (typeof key !== "string" || key.length === 0) continue;
                obj[key] = c < rows[r].length ? rows[r][c] : "";
            }
            result.push(obj);
        }
        return result;
    }

    return rows;
}

/**
 * Encode rows into a CSV string.
 *
 * Values containing the separator, quote, CR, or LF are auto-quoted;
 * embedded quotes are doubled (RFC 4180).
 *
 * @param {Array<Array>|Array<Object>} rows  When `opts.headers=true`, rows
 *   are objects (keys become the header row). Otherwise arrays.
 * @param {Object} [opts]
 * @param {boolean} [opts.headers=false]
 * @param {string}  [opts.separator=","]
 * @param {string}  [opts.quote='"']
 * @param {boolean} [opts.sanitizeFormulas=false]  Prefix a "'" to any cell
 *   beginning with = + - @ (or a leading tab/CR) to neutralize spreadsheet
 *   formula/DDE injection when the export is opened in Excel/Sheets. Off by
 *   default (it prepends a character to affected cells). Lua-parity alias:
 *   `sanitize_formulas`.
 * @returns {string}  CSV text (LF line endings).
 */
function encode(rows, opts) {
    if (!Array.isArray(rows) || rows.length === 0) return "";

    const sep = (opts && opts.separator) || ",";
    const quo = (opts && opts.quote) || '"';
    const useHeaders = !!(opts && opts.headers);
    // Opt-in CSV formula-injection defense (Lua-parity alias: sanitize_formulas):
    // prefix a "'" to any cell beginning with = + - @ (or a leading tab/CR) so a
    // spreadsheet app treats it as text, not a formula/DDE. Off by default.
    const sanitize = !!(opts && (opts.sanitizeFormulas ?? opts.sanitize_formulas));
    const doubledQuote = quo + quo;

    function needsQuoting(value) {
        for (let i = 0; i < value.length; i++) {
            const ch = value[i];
            if (ch === sep || ch === quo || ch === "\n" || ch === "\r")
                return true;
        }
        return false;
    }

    function quoteField(value) {
        let s = String(value);
        if (sanitize && s.length > 0) {
            const c = s[0];
            if (c === "=" || c === "+" || c === "-" || c === "@" ||
                c === "\t" || c === "\r")
                s = "'" + s;
        }
        if (needsQuoting(s)) {
            const parts = [];
            for (let i = 0; i < s.length; i++) {
                if (s[i] === quo) parts.push(doubledQuote);
                else parts.push(s[i]);
            }
            return quo + parts.join("") + quo;
        }
        return s;
    }

    function encodeRow(fields) {
        const parts = [];
        for (let i = 0; i < fields.length; i++) {
            parts.push(quoteField(fields[i] === null || fields[i] === undefined ? "" : fields[i]));
        }
        return parts.join(sep);
    }

    let out = "";

    if (useHeaders) {
        // rows is array of objects; derive headers from first object keys
        if (typeof rows[0] !== "object" || rows[0] === null) return "";

        const headers = Object.keys(rows[0]);
        out += encodeRow(headers) + "\n";

        for (let r = 0; r < rows.length; r++) {
            const values = [];
            for (let c = 0; c < headers.length; c++) {
                values.push(rows[r][headers[c]]);
            }
            out += encodeRow(values) + "\n";
        }
    } else {
        // rows is array of arrays
        for (let r = 0; r < rows.length; r++) {
            out += encodeRow(rows[r]) + "\n";
        }
    }

    return out;
}

const csv = { parse, encode };
export { csv };
